### PR TITLE
List btrfs snapshots separately in the subvolume detail page

### PR DESCRIPTION
**Pull request by jelly**
This PR identifies snapshots and lists them under the subvolume in a separate card.

Open questions:
* [ ] Should we have actions on snapshots when they are listed in the crossrefs table? At least "Delete" seems appropriate.
* [ ] In the detail view, we now list snapshots and subvolumes both, We should probably hide snapshots from the subvolume list?
* [ ] Do we show snapshots are "btrfs snapshot" instead of "btrfs subvolume".

---

# Show snapshots of a subvolume in the detail view

image

Snapshot origin is shown in snapshot detail view

image

### CODE REVIEW

**Comment by jelly** [comment]
@mvollmer I thought a bit on how we could implement filtering:

We could come up with a system where we don't have to hardcode the "filters" in but we could register filters for pages which want that.

```
register_filter(display_name, property_name)
```

This registers a filter which lives in `pages.jsx`, then we use these filter with `property_name` and `filter_properties.`

Alternatively, we just hardcode filters:

```
PAGE_FILTER_SNAPSHOT = 'snapshot'
```

Page/Card can have `filter_properties` which contains the key we are interested for with the value. State can come later.

I'm not thinking of introducing something like a [filter component](https://www.patternfly.org/patterns/filters#checkbox-select). But maybe it is interesting to have something usable.

**Comment by mvollmer** [comment]
> @mvollmer I thought a bit on how we could implement filtering:

I think this is a bit too much machinery. Let's just do what we need, and nothing more.

**Comment by jelly** [comment]
Alright, after the discussion today in standup:

* First pass, show snapshots under a subvolume
* Second pass, hide snapshots

**Comment by garrett** [comment]
@jelly: Yes, I think that's what we want. Good, simple summary.

 I think we (you and/or @mvollmer and I) agreed on that a while ago somewhere? Perhaps in a video call? Perhaps in an issue? I don't remember where, but if we all agree on that, then that's great.

To restart: I definitely do not want them on the overview (and never have) and I think the details pages are the right places for the snapshots.

**Comment by mvollmer** [comment]
So the listing of snapshots on the page of their origin in a crossref table works. Let's nail down when to hide snapshots in the regular child tables.

What about this: snapshots are only included in a child table when they are on the top-level. Otherwise they are replaced with a row that summarizes them "(20 snapshots)". Clicking on that row will go to the parent of those snapshots.

Here is a made-up example (_without_ that rule applied):

image

Here, all the home-123 and images-123 subvolumes are snapshots. They would be omitted from the table and replaced with "(13 snapshots)".

But when going to the "snapshots" subvolume, they would be listed like they are now, because they are on the top-level in that table:

image

**Comment by jelly** [comment]
> So the listing of snapshots on the page of their origin in a crossref table works. Let's nail down when to hide snapshots in the regular child tables.

Good point.... I forgot about that...

So on the overview they need to go away, that is a requirement from @garrett 
 
> What about this: snapshots are only included in a child table when they are on the top-level. Otherwise they are replaced with a row that summarizes them "(20 snapshots)". Clicking on that row will go to the parent of those snapshots.

What view is this? The filesystem detail page? Can we differentiate the overview and filesystem detail page easily?

> Here is a made-up example (_without_ that rule applied):
> 
> image
> 
> Here, all the home-123 and images-123 subvolumes are snapshots. They would be omitted from the table and replaced with "(13 snapshots)".

Hmm feels a bit magical to me.
 
> But when going to the "snapshots" subvolume, they would be listed like they are now, because they are on the top-level in that table:
> 
> image

But then you wouldn't know it is a special subvolume would you? Better asked, how would we implement this:

```
for subvolume in parent_dir.subvolumes:
       if subvolume.parent__uid == parent_dir.uuid
            show()
      else
           hide()
```

The problem is a bit that BTRFS is very lenient and you can make everything everywhere without having to give it logical names.

> And of course they would also be listed on the page of their origin:
> 
> image

Agreed.

> In addition the page for a snapshot subvolume should have a link back to its origin:
> 
> image
> 
> Heere the could be a description row with "Snapshot origin data/home".

Stretch goal, but seems nice!

**Comment by mvollmer** [comment]
> Better asked, how would we implement this:
> 
> ```
> for subvolume in parent_dir.subvolumes:
>        if subvolume.parent__uid == parent_dir.uuid
>             show()
>       else
>            hide()
> ```

So parent_uuid is the origin of the snaphot subvolume, and parent_dir is the parent subvolume, right?  In my example, "snapshots" is the parent of "home-1234", and "home" is the origin of "home-1234" (since it was created by "btrfs subvolume snapshot /run/butter/data/home /run/butter/data/snapshots/home-1234").

So your code above would not hide any snapshots in my example, since no snapshot subvolume has the subvolume as both parent and origin. It would only hide snapshots that are directly inside their origins, like the one created by "btrfs subvolume snapshot /run/butter/data/home /run/butter/data/home/snapshot-1".

In my understanding, we want to get rid of snapshots in our tables because there are too many of them and they are usually not interesting.

**Comment by mvollmer** [comment]
> > What about this: snapshots are only included in a child table when they are on the top-level. Otherwise they are replaced with a row that summarizes them "(20 snapshots)". Clicking on that row will go to the parent of those snapshots.
> 
> What view is this? The filesystem detail page? Can we differentiate the overview and filesystem detail page easily?

That rule would apply to any ChildrenTable, thus to each page, overview and details.

We can distinguish overview  and details pages and we do that already: only the overview has icons, for example), But I propose we don't do that when deciding whether to hide snapshots. If they are distracting on the overview, they are also distracting elsewhere, no? And with the summary line, you can easily find the place that lists them all.

**Comment by mvollmer** [comment]
Should we have actions on snapshots when they are listed in the crossrefs table? At least "Delete" seems appropriate.

**Comment by jelly** [comment]
> > Better asked, how would we implement this:
> > ```
> > for subvolume in parent_dir.subvolumes:
> >        if subvolume.parent__uid == parent_dir.uuid
> >             show()
> >       else
> >            hide()
> > ```
> 
> So parent_uuid is the origin of the snaphot subvolume, and parent_dir is the parent subvolume, right? In my example, "snapshots" is the parent of "home-1234", and "home" is the origin of "home-1234" (since it was created by "btrfs subvolume snapshot /run/butter/data/home /run/butter/data/snapshots/home-1234").
> 
> So your code above would not hide any snapshots in my example, since no snapshot subvolume has the subvolume as both parent and origin. It would only hide snapshots that are directly inside their origins, like the one created by "btrfs subvolume snapshot /run/butter/data/home /run/butter/data/home/snapshot-1".

No this does indeed not work...

```
[root@fedora-40-127-0-0-2-2201 ~]# btrfs subvolume list -apuq /
ID 256 gen 838 parent 5 top level 5 parent_uuid -                                    uuid 13c51cda-d57a-0a45-a808-1785eca3bc28 path <FS_TREE>/root
ID 257 gen 838 parent 5 top level 5 parent_uuid -                                    uuid 9d64752d-1de6-2849-baac-696ce79f3a7b path <FS_TREE>/home
ID 258 gen 840 parent 5 top level 5 parent_uuid -                                    uuid bd3edcf1-e9b5-654a-96ec-6e62ee17d924 path <FS_TREE>/var
ID 259 gen 78 parent 258 top level 258 parent_uuid -                                    uuid b3735d93-91e0-1046-98f7-e377fafc1b3c path <FS_TREE>/var/lib/machines
ID 260 gen 836 parent 256 top level 256 parent_uuid -                                    uuid 6b5445f7-2d40-5c45-b914-9eedd47dd5cd path root/snapshots
ID 261 gen 836 parent 260 top level 260 parent_uuid 9d64752d-1de6-2849-baac-696ce79f3a7b uuid cf9b06e1-6141-0b42-b61c-5d2c920e88ab path <FS_TREE>/root/snapshots/today
```

> In my understanding, we want to get rid of snapshots in our tables because there are too many of them and they are usually not interesting. Isn't that the goal?

Yes, but you argued for showing them?

**Comment by jelly** [comment]
> Should we have actions on snapshots when they are listed in the crossrefs table? At least "Delete" seems appropriate.

Sounds reasonable.

**Comment by mvollmer** [comment]
> > In my understanding, we want to get rid of snapshots in our tables because there are too many of them and they are usually not interesting. Isn't that the goal?
> 
> Yes, but you argued for showing them?

Yes, but sparingly. Snapshots are also subvolumes and we need to offer all the functionality of subvolumes for them, somewhere. Maybe people want to mount them, or copy them back to their origin.

**Comment by jelly** [comment]
Conclusion from standup with @mvollmer "Let's merge this and figure out filtering/hiding subvolumes later.

File: pkg/storaged/btrfs/subvolume.jsx [hunk 1]
```diff
@@ -395,6 +396,14 @@ function make_btrfs_subvolume_page(parent, volume, subvol, path_prefix, subvols)
         props: { subvol, mount_point, mismount_warning, block, fstab_config, forced_options },
         actions,
     });
+
+    if (subvol.id !== 5 && subvol.parent_uuid !== null)
+        register_crossref({
+            key: subvol.parent_uuid,
+            card,
+            size: mounted && <StorageUsageBar stats={use} short />,
```
Comment (cockpituous) [inline pkg/storaged/btrfs/subvolume.jsx:404]: This added line is not executed by any test.

Comment (cockpituous) [inline pkg/storaged/btrfs/subvolume.jsx:404]: This added line is not executed by any test.

Comment (cockpituous) [inline pkg/storaged/btrfs/subvolume.jsx:404]: This added line is not executed by any test.

**Comment by jelly** [comment]
@martinpitt do you mind taking the review over?

**Comment by jelly** [comment]
Alternative idea, we have one list of subvolumes:

* Show subvolumes and snapshots different, either with a Label 
* Allow filtering of the table

File: pkg/storaged/btrfs/subvolume.jsx [hunk 2]
```diff
@@ -392,9 +403,17 @@ function make_btrfs_subvolume_page(parent, volume, subvol, path_prefix, subvols)
...
-        props: { subvol, mount_point, mismount_warning, block, fstab_config, forced_options },
+        props: { subvol, snapshot_origin, mount_point, mismount_warning, block, fstab_config, forced_options },
         actions,
     });
+
+    if (subvol.id !== 5 && subvol.parent_uuid !== null)
+        register_crossref({
+            key: subvol.parent_uuid,
+            card,
+            size: mounted && <StorageUsageBar stats={use} short />,
```
Comment (cockpituous) [inline pkg/storaged/btrfs/subvolume.jsx:414]: This added line is not executed by any test.

File: pkg/storaged/btrfs/subvolume.jsx [hunk 3]
```diff
@@ -413,6 +434,9 @@ const BtrfsSubvolumeCard = ({ card, subvol, mismount_warning, block, fstab_confi
                 <DescriptionList className="pf-m-horizontal-on-sm">
                     <StorageDescription title={_("Name")} value={subvol.pathname} />
                     <StorageDescription title={_("ID")} value={subvol.id} />
+                    {snapshot_origin !== null &&
+                    <StorageDescription title={_("Snapshot origin")} value={snapshot_origin.pathname} />
```
Comment (cockpituous) [inline pkg/storaged/btrfs/subvolume.jsx:438]: This added line is not executed by any test.

Comment (mvollmer) [inline pkg/storaged/btrfs/subvolume.jsx:438]: This should be a link, no?

File: pkg/storaged/btrfs/subvolume.jsx [hunk 4]
```diff
@@ -392,9 +403,17 @@ function make_btrfs_subvolume_page(parent, volume, subvol, path_prefix, subvols)
         location: mp_text,
         component: BtrfsSubvolumeCard,
         has_warning: !!mismount_warning,
-        props: { subvol, mount_point, mismount_warning, block, fstab_config, forced_options },
+        props: { subvol, snapshot_origin, mount_point, mismount_warning, block, fstab_config, forced_options },
         actions,
     });
+
+    if (subvol.id !== 5 && subvol.parent_uuid !== null)
+        register_crossref({
```
Comment (mvollmer) [inline pkg/storaged/btrfs/subvolume.jsx:411]: Let's add the "Delete" action here.

**Review by mvollmer — CHANGES_REQUESTED** [review]

**Review by mvollmer — APPROVED** [review]
Thanks!

File: pkg/storaged/btrfs/subvolume.jsx [hunk 5]
```diff
@@ -392,9 +404,17 @@ function make_btrfs_subvolume_page(parent, volume, subvol, path_prefix, subvols)
...
-        props: { subvol, mount_point, mismount_warning, block, fstab_config, forced_options },
+        props: { volume, subvol, snapshot_origin, mount_point, mismount_warning, block, fstab_config, forced_options },
         actions,
     });
+
+    if (subvol.id !== 5 && subvol.parent_uuid !== null)
+        register_crossref({
+            key: subvol.parent_uuid,
+            card,
+            size: mounted && <StorageUsageBar stats={use} short />,
```
Comment (cockpituous) [inline pkg/storaged/btrfs/subvolume.jsx:415]: This added line is not executed by any test.

**Comment by 2nv2u** [comment]
Can you add the timestamp of the snapshots to the overview and detail page?
Not very useful for subvolumes, but it means a lot for snapshots.

**Comment by jelly** [comment]
> Can you add the timestamp of the snapshots to the overview and detail page? Not very useful for subvolumes, but it means a lot for snapshots.

subvolume list does not expose the time the snapshot is made. If you want this implement I think we have to stat the snapshot directory? Either way it would be useful to have a new issue if you want this feature request researched.

**Comment by 2nv2u** [comment]
Yes that's correct, it only show in the stats of the snap shot itself `btrfs subvol show {snapshot}`, I'll create a ticket for this.

**Pull request merged**